### PR TITLE
machinst x64: check SSE requirements for instructions

### PR DESCRIPTION
**Pull request by bnjbvr**
This builds on top of #2255 (two commits), and adds a dynamic check that every time we emit an instruction that uses SSE features, the features were enabled by the configuration. This avoids cases where we could use e.g. SSE4.1 on a machine that doesn't have it.

The way to pass the ISA flags is a bit tricky; please see commit message for the explanation of why it was done this way. Happy to try a different way if anybody has better suggestions!

### CODE REVIEW

File: cranelift/codegen/src/isa/x64/inst/mod.rs [hunk 1]
```diff
@@ -489,6 +488,70 @@ pub(crate) fn low32_will_sign_extend_to_64(x: u64) -> bool {
     xs == ((xs << 32) >> 32)
 }
 
+impl Inst {
+    fn iset_requirement(&self) -> Option<InstructionSet> {
```
Comment (cfallin) [inline cranelift/codegen/src/isa/x64/inst/mod.rs:492]: Could we rename this to `isa_requirement()`? I think "ISA" is a more standard term than "iset" as an abbreviation.

File: cranelift/codegen/src/isa/x64/inst/mod.rs [hunk 2]
```diff
@@ -489,6 +488,70 @@ pub(crate) fn low32_will_sign_extend_to_64(x: u64) -> bool {
...
+        match self {
+            // These instructions are part of SSE2, which is a basic requirement in Cranelift, and
+            // don't have to be checked.
```
Comment (cfallin) [inline cranelift/codegen/src/isa/x64/inst/mod.rs:495]: Are all of the below SSE2, e.g. `JmpIf`? (Perhaps they have some options or encodings that are?)

(To be clear, I like this function and the explicitness of this `match`; I'm just curious about the comment is all.)

**Review by cfallin — APPROVED** [review]
Very nice, thanks! I think the `MachInstEmitInfo` approach is reasonable; it's useful to have the ability to hold some constant configuration outside of individual instructions.

I do wonder if we should document any invariants related to this: for example, is it permissible to make changes in the `EmitInfo` and expect different code to be emitted, or (as I suspect instead) is the specific vcode lowering somehow tied to an `EmitInfo`? If the latter, would it make sense to hold the `EmitInfo` in the `VCode` container itself, and plumb it into the instruction emission from there?

Approving either way, but happy to take another look if you refactor in that direction.

File: cranelift/codegen/src/isa/x64/inst/mod.rs (hunk 2 above)
Comment (bnjbvr) [reply cranelift/codegen/src/isa/x64/inst/mod.rs:495]: Yep, I just checked the ones for which i had a few doubts:

- the conversions from int to float/doubles (and reciprocally) are SSE (for floats) and SSE2 (for doubles)
- JmpIf is just a plain old conditional jump, (e.g. `jo`), so nothing to fret about.

**Comment by bnjbvr** [comment]
> would it make sense to hold the EmitInfo in the VCode container itself, and plumb it into the instruction emission from there?

I'm happy to do this, but that would break the symmetry with `EmitState`... which could also live in Vcode in fact. Maybe also moving EmitState would help reducing the traits complexity... Any preference here? (Happy to chat about this on Zulip)

**Comment by cfallin** [comment]
I suppose I see the difference between `EmitState` and `EmitInfo` as: the vcode should be emittable multiple times (the instructions are immutable during emit), so we want `EmitState` to be an external thing -- it's transient, per emission; while `EmitInfo` encapsulates some information that is tied to the lowering and should not be changed, so its lifetime should be as long as the VCode. (And given that it shouldn't change, there is really only one "correct value" to use when emitting a given VCode body, so it doesn't seem right to require it as a parameter.) So from that perspective, the asymmetry doesn't bother me too much. That said, I'm happy to go with your original design as well if you prefer it :-)

**Comment by bnjbvr** [comment]
That's a good argument, thanks! I hadn't thought about the fact that we could emit several times (I wonder what such a use case would be though, since an emitter who'd like to emit in several places would emit in a buffer, and then would carry a cheap copy rather than the full-blown emission.) I'll rework my patch.

**Comment by bnjbvr** [comment]
@cfallin Updated the patch, please let me know if you want to take another look!

**Review by cfallin — APPROVED** [review]
Looks good!

**Pull request merged**